src/cpp/src/continuous_batching_pipeline.cpp

```diff
@@ -65,6 +65,9 @@ class ContinuousBatchingPipeline::Impl {
         while (requests_iterator != m_requests.end()) {
             const auto& request = *requests_iterator;
             if(request->has_finished() || request->out_of_memory() || request->handle_dropped()) {
+                // Notify the last time even if there will be no results
+                // This causes read_all() to unblock in all situations
+                request->notify_handle();
```

Comment:
I don't think we can do it that way. The idea is that we should have only one notification per step() (if any). Here we add another call and when called in such circumstances, `notify_handle()` will always send tokens to handle. So we can end up  sending the same results twice. For example when generation finishes it notifies handle in sampler and then here in cleanup method. In the streaming scenario, if generation handle doesn't read() token and check status between these calls it will read last token twice.

Reply:
I have removed notify_handle call in case there is out of memory error. Now it is not duplicated, those will be notified in _free_non_running_requests.